README
NAME
    RT-Extension-CustomiseMenus - enables easy modification of RT menus

DESCRIPTION
    This extension for RT allows a administrator to easily add/remove/rename
    the RT menus.

RT VERSION
    Works with RT 4.4

INSTALLATION
    perl Makefile.PL
    make
    make install
        May need root permissions

    Edit your /opt/rt4/etc/RT_SiteConfig.pm
        Adding this line:

            Plugin('RT::Extension::CustomiseMenus');

    Clear your mason cache
            rm -rf /opt/rt4/var/mason_data/obj

    Restart your webserver

CONFIGURATION
    A RT menu can be manipluated by adding a simple configuration array of
    hashes as shown below:

        Set($PrivilegedMenus, [
                    {
                            'id'     => 'corporate',
                            'action' => 'add',
                            'path'   => 'http://bestpractical.com',
                            'title'  => 'Corporate',
                            'menu'   => 'main',
                            'after'  => 'home',
                    },
                    {
                            'id'     => 'wiki',
                            'action' => 'add',
                            'path'   => 'http://wiki.bestpractical.com',
                            'title'  => 'Wiki',
                            'menu'   => 'main',
                            'parent' => 'corporate',
                    },
                    {
                            'id'     => 'newitem',
                            'action' => 'add',
                            'path'   => '/new/thing/here',
                            'title'  => 'New Action',
                            'menu'   => 'tabs',
                            'parent' => 'actions',
                    },
                    {
                            'id'     => 'tools',
                            'action' => 'remove',
                            'menu'   => 'main',
                    },
        ]);

        Set($SelfServiceMenus, [

        ]);

AUTHOR
    Martin Wheldon <martin.wheldon@greenhills-it.co.uk>

BUGS
    All bugs should be reported via email to

        L<bug-RT-Extension-CustomiseMenus@greenhills-it.co.uk|mailto:bug-RT-Extension-CustomiseMenus@greenhills-it.co.uk>

LICENSE AND COPYRIGHT
    This software is Copyright (c) 2017 by Greenhills IT Ltd.

    This is free software, licensed under:

      The GNU General Public License, Version 2, June 1991



Changes
Revision history for RT-Extension-CustomiseMenus

0.01 [22-05-2017]
 - Initial version


inc/Module/Install/RTx.pm
#line 1
package Module::Install::RTx;

use 5.008;
use strict;
use warnings;
no warnings 'once';

use Module::Install::Base;
use base 'Module::Install::Base';
our $VERSION = '0.38';

use FindBin;
use File::Glob     ();
use File::Basename ();

my @DIRS = qw(etc lib html static bin sbin po var);
my @INDEX_DIRS = qw(lib bin sbin);

sub RTx {
    my ( $self, $name, $extra_args ) = @_;
    $extra_args ||= {};

    # Set up names
    my $fname = $name;
    $fname =~ s!-!/!g;

    $self->name( $name )
        unless $self->name;
    $self->all_from( "lib/$fname.pm" )
        unless $self->version;
    $self->abstract("$name Extension")
        unless $self->abstract;
    unless ( $extra_args->{no_readme_generation} ) {
        $self->readme_from( "lib/$fname.pm",
                            { options => [ quotes => "none" ] } );
    }
    $self->add_metadata("x_module_install_rtx_version", $VERSION );

    my $installdirs = $ENV{INSTALLDIRS};
    for ( @ARGV ) {
        if ( /INSTALLDIRS=(.*)/ ) {
            $installdirs = $1;
        }
    }

    # Try to find RT.pm
    my @prefixes = qw( /opt /usr/local /home /usr /sw /usr/share/request-tracker4);
    $ENV{RTHOME} =~ s{/RT\.pm$}{} if defined $ENV{RTHOME};
    $ENV{RTHOME} =~ s{/lib/?$}{}  if defined $ENV{RTHOME};
    my @try = $ENV{RTHOME} ? ($ENV{RTHOME}, "$ENV{RTHOME}/lib") : ();
    while (1) {
        my @look = @INC;
        unshift @look, grep {defined and -d $_} @try;
        push @look, grep {defined and -d $_}
            map { ( "$_/rt4/lib", "$_/lib/rt4", "$_/lib" ) } @prefixes;
        last if eval {local @INC = @look; require RT; $RT::LocalLibPath};

        warn
            "Cannot find the location of RT.pm that defines \$RT::LocalPath in: @look\n";
        my $given = $self->prompt("Path to directory containing your RT.pm:") or exit;
        $given =~ s{/RT\.pm$}{};
        $given =~ s{/lib/?$}{};
        @try = ($given, "$given/lib");
    }

    print "Using RT configuration from $INC{'RT.pm'}:\n";

    my $local_lib_path = $RT::LocalLibPath;
    unshift @INC, $local_lib_path;
    my $lib_path = File::Basename::dirname( $INC{'RT.pm'} );
    unshift @INC, $lib_path;

    # Set a baseline minimum version
    unless ( $extra_args->{deprecated_rt} ) {
        $self->requires_rt('4.0.0');
    }

    # Installation locations
    my %path;
    my $plugin_path;
    if ( $installdirs && $installdirs eq 'vendor' ) {
        $plugin_path = $RT::PluginPath;
    } else {
        $plugin_path = $RT::LocalPluginPath;
    }
    $path{$_} = $plugin_path . "/$name/$_"
        foreach @DIRS;

    # Copy RT 4.2.0 static files into NoAuth; insufficient for
    # images, but good enough for css and js.
    $path{static} = "$path{html}/NoAuth/"
        unless $RT::StaticPath;

    # Delete the ones we don't need
    delete $path{$_} for grep {not -d "$FindBin::Bin/$_"} keys %path;

    my %index = map { $_ => 1 } @INDEX_DIRS;
    $self->no_index( directory => $_ ) foreach grep !$index{$_}, @DIRS;

    my $args = join ', ', map "q($_)", map { ($_, "\$(DESTDIR)$path{$_}") }
        sort keys %path;

    printf "%-10s => %s\n", $_, $path{$_} for sort keys %path;

    if ( my @dirs = map { ( -D => $_ ) } grep $path{$_}, qw(bin html sbin etc) ) {
        my @po = map { ( -o => $_ ) }
            grep -f,
            File::Glob::bsd_glob("po/*.po");
        $self->postamble(<< ".") if @po;
lexicons ::
\t\$(NOECHO) \$(PERL) -MLocale::Maketext::Extract::Run=xgettext -e \"xgettext(qw(@dirs @po))\"
.
    }

    $self->include('Module::Install::RTx::Runtime') if $self->admin;
    $self->include_deps( 'YAML::Tiny', 0 ) if $self->admin;
    my $postamble = << ".";
install ::
\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxPlugin()"
\t\$(NOECHO) \$(PERL) -MExtUtils::Install -e \"install({$args})\"
.

    if ( $path{var} and -d $RT::MasonDataDir ) {
        my ( $uid, $gid ) = ( stat($RT::MasonDataDir) )[ 4, 5 ];
        $postamble .= << ".";
\t\$(NOECHO) chown -R $uid:$gid $path{var}
.
    }

    my %has_etc;
    if ( File::Glob::bsd_glob("$FindBin::Bin/etc/schema.*") ) {
        $has_etc{schema}++;
    }
    if ( File::Glob::bsd_glob("$FindBin::Bin/etc/acl.*") ) {
        $has_etc{acl}++;
    }
    if ( -e 'etc/initialdata' ) { $has_etc{initialdata}++; }
    if ( grep { /\d+\.\d+\.\d+.*$/ } glob('etc/upgrade/*.*.*') ) {
        $has_etc{upgrade}++;
    }

    my %has_patches;
    if ( -d 'patches' ) {
        $has_patches{patches}++;
    }

    $self->postamble("$postamble\n");
    if ( $path{lib} ) {
        $self->makemaker_args( INSTALLSITELIB => $path{'lib'} );
        $self->makemaker_args( INSTALLARCHLIB => $path{'lib'} );
        $self->makemaker_args( INSTALLVENDORLIB => $path{'lib'} )
    } else {
        $self->makemaker_args( PM => { "" => "" }, );
    }

    $self->makemaker_args( INSTALLSITEMAN1DIR => "$RT::LocalPath/man/man1" );
    $self->makemaker_args( INSTALLSITEMAN3DIR => "$RT::LocalPath/man/man3" );
    $self->makemaker_args( INSTALLSITEARCH => "$RT::LocalPath/man" );

    # INSTALLDIRS=vendor should install manpages into /usr/share/man.
    # That is the default path in most distributions. Need input from
    # Redhat, Centos etc.
    $self->makemaker_args( INSTALLVENDORMAN1DIR => "/usr/share/man/man1" );
    $self->makemaker_args( INSTALLVENDORMAN3DIR => "/usr/share/man/man3" );
    $self->makemaker_args( INSTALLVENDORARCH => "/usr/share/man" );

    if (%has_etc) {
        print "For first-time installation, type 'make initdb'.\n";
        my $initdb = '';
        $initdb .= <<"." if $has_etc{schema};
\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxDatabase(qw(schema \$(NAME) \$(VERSION)))"
.
        $initdb .= <<"." if $has_etc{acl};
\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxDatabase(qw(acl \$(NAME) \$(VERSION)))"
.
        $initdb .= <<"." if $has_etc{initialdata};
\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxDatabase(qw(insert \$(NAME) \$(VERSION)))"
.
        $self->postamble("initdb ::\n$initdb\n");
        $self->postamble("initialize-database ::\n$initdb\n");
        if ($has_etc{upgrade}) {
            print "To upgrade from a previous version of this extension, use 'make upgrade-database'\n";
            my $upgradedb = qq|\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxDatabase(qw(upgrade \$(NAME) \$(VERSION)))"\n|;
            $self->postamble("upgrade-database ::\n$upgradedb\n");
            $self->postamble("upgradedb ::\n$upgradedb\n");
        }
    }

    if (%has_patches) {
        print "For first-time installation, type 'make patch'.\n";
        my $patch = qq|\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxPatch(qw(/usr/bin/patch apply ./patches))"\n|;
        $self->postamble("patch ::\n$patch\n");

        print "For patch removal, type 'make unpatch'.\n";
        my $unpatch = qq|\t\$(NOECHO) \$(PERL) -Ilib -I"$local_lib_path" -I"$lib_path" -Iinc -MModule::Install::RTx::Runtime -e"RTxPatch(qw(/usr/bin/patch remove ./patches))"\n|;
        $self->postamble("unpatch ::\n$unpatch\n");
    }

}

sub requires_rt {
    my ($self,$version) = @_;

    _load_rt_handle();

    if ($self->is_admin) {
        $self->add_metadata("x_requires_rt", $version);
        my @sorted = sort RT::Handle::cmp_version $version,'4.0.0';
        $self->perl_version('5.008003') if $sorted[0] eq '4.0.0'
            and (not $self->perl_version or '5.008003' > $self->perl_version);
        @sorted = sort RT::Handle::cmp_version $version,'4.2.0';
        $self->perl_version('5.010001') if $sorted[0] eq '4.2.0'
            and (not $self->perl_version or '5.010001' > $self->perl_version);
    }

    # if we're exactly the same version as what we want, silently return
    return if ($version eq $RT::VERSION);

    my @sorted = sort RT::Handle::cmp_version $version,$RT::VERSION;

    if ($sorted[-1] eq $version) {
        die <<"EOT";

**** Error: This extension requires RT $version. Your installed version
            of RT ($RT::VERSION) is too old.

EOT
    }
}

sub requires_rt_plugin {
    my $self = shift;
    my ( $plugin ) = @_;

    if ($self->is_admin) {
        my $plugins = $self->Meta->{values}{"x_requires_rt_plugins"} || [];
        push @{$plugins}, $plugin;
        $self->add_metadata("x_requires_rt_plugins", $plugins);
    }

    my $path = $plugin;
    $path =~ s{\:\:}{-}g;
    $path = "$RT::LocalPluginPath/$path/lib";
    if ( -e $path ) {
        unshift @INC, $path;
    } else {
        my $name = $self->name;
        warn <<"EOT";

**** Warning: $name requires that the $plugin plugin be installed and
              enabled; it does not appear to be installed.

EOT
    }
    $self->requires(@_);
}

sub rt_too_new {
    my ($self,$version,$msg) = @_;
    my $name = $self->name;
    $msg ||= <<EOT;

**** Error: Your installed version of RT (%s) is too new; this extension
            only works with versions older than %s.

EOT
    $self->add_metadata("x_rt_too_new", $version) if $self->is_admin;

    _load_rt_handle();
    my @sorted = sort RT::Handle::cmp_version $version,$RT::VERSION;

    if ($sorted[0] eq $version) {
        die sprintf($msg,$RT::VERSION,$version);
    }
}

# RT::Handle runs FinalizeDatabaseType which calls RT->Config->Get
# On 3.8, this dies.  On 4.0/4.2 ->Config transparently runs LoadConfig.
# LoadConfig requires being able to read RT_SiteConfig.pm (root) so we'd
# like to avoid pushing that on users.
# Fake up just enough Config to let FinalizeDatabaseType finish, and
# anyone later calling LoadConfig will overwrite our shenanigans.
sub _load_rt_handle {
    unless ($RT::Config) {
        require RT::Config;
        $RT::Config = RT::Config->new;
        RT->Config->Set('DatabaseType','mysql');
    }
    require RT::Handle;
}

1;

__END__

#line 446


inc/Module/Install/RTx/Runtime.pm
#line 1
package Module::Install::RTx::Runtime;

use base 'Exporter';
our @EXPORT = qw/RTxDatabase RTxPatch RTxPlugin/;

use strict;
use File::Basename ();
use POSIX qw (WIFEXITED);
use File::Find;
use Carp;
use Cwd;


# Switch
my $dispatch  ={
    '>' => \&_greater_than,
    '<' => \&_less_than,
    '-' => \&_in_range,
    '=' => \&_equal,
};

sub _in_range {
    my ($rt_ver, $dir) = @_;

    # Get our lower and upper versions
    my ($start_ver, $end_ver) = $dir =~ m/([\d\.]+)-([\d\.]+)/xms;

    # Check if rt_ver is between range of version values
    return 1 if &_greater_than($rt_ver, $start_ver)
                and &_less_than($rt_ver, $end_ver);
    return 1 if &_equal($rt_ver, $start_ver);
    return 1 if &_equal($rt_ver, $end_ver);
    return 0;
}

sub _greater_than {
    my ($rt_ver, $patch_ver) = @_;

    $patch_ver =~ s/\A>//xms;
    $patch_ver = _convert_version($patch_ver);
    $rt_ver = _convert_version($rt_ver);

    # Return TRUE if we are greater than the patch
    return 1 if $rt_ver > $patch_ver;
    return 0;
}

sub _less_than {
    my ($rt_ver, $patch_ver) = @_;

    $patch_ver =~ s/\A<//xms;
    $patch_ver = _convert_version($patch_ver);
    $rt_ver = _convert_version($rt_ver);

    # Return TRUE if we are less than the patch
    return 1 if $rt_ver < $patch_ver;
    return 0;
}

sub _equal {
    my ($rt_ver, $patch_ver) = @_;

    $patch_ver = _convert_version($patch_ver);
    $rt_ver = _convert_version($rt_ver);

    # Check if the versions are equal
    return 1 if $rt_ver == $patch_ver;
    return 0;
}

sub _convert_version {
    my $version = shift @_;
    my @nums;
    my $num;

    # Pass version straight back if no "." is found in it
    # else convert to a float in the format \d+.\d+
    if ($version =~ m/\./xms) {
        @nums = split(/\./, $version);
        $num = shift @nums;
        $num .= '.';
        $num .= join('0', @nums);
        return $num;
    }
    else {
        return $version;
    }
}

sub _required_patch {
    my ($path, $version) = @_;

    # We are pretty safe doing this as the ./patches path is hardcoded
    my @dirs = split('/', $path);
    my $dir = $dirs[2];

    # Check range of versions
    return $dispatch->{'-'}($version, $dir) if $dir =~ m/[\d.]+.\d-\d.[\d.]+/xms;
    # Check greater than
    return $dispatch->{'>'}($version, $dir) if $dir =~ m/\A>[\d.]+\z/xms;
    # Check less than
    return $dispatch->{'<'}($version, $dir) if $dir =~ m/\A<[\d.]+\z/xms;
    # Check if equal
    return $dispatch->{'='}($version, $dir) if $dir =~ m/\A\d.\d+\z/xms;

    # If none of the above match then just apply the patch(s) regardless
    return 1;
}

sub _rt_runtime_load {
    require RT;

    eval { RT::LoadConfig(); };
    if (my $err = $@) {
        die $err unless $err =~ /^RT couldn't load RT config file/m;
        my $warn = <<EOT;
This usually means that your current user cannot read the file.  You
will likely need to run this installation step as root, or some user
with more permissions.
EOT
        $err =~ s/This usually means.*/$warn/s;
        die $err;
    }
}

sub RTxDatabase {
    my ($action, $name, $version) = @_;

    _rt_runtime_load();

    require RT::System;
    my $has_upgrade = RT::System->can('AddUpgradeHistory');

    my $lib_path = File::Basename::dirname($INC{'RT.pm'});
    my @args = (
        "-Ilib",
        "-I$RT::LocalLibPath",
        "-I$lib_path",
        "$RT::SbinPath/rt-setup-database",
        "--action"      => $action,
        ($action eq 'upgrade' ? () : ("--datadir"     => "etc")),
        (($action eq 'insert') ? ("--datafile"    => "etc/initialdata") : ()),
        "--dba"         => $RT::DatabaseAdmin || $RT::DatabaseUser,
        "--prompt-for-dba-password" => '',
        ($has_upgrade ? ("--package" => $name, "--ext-version" => $version) : ()),
    );
    # If we're upgrading against an RT which isn't at least 4.2 (has
    # AddUpgradeHistory) then pass --package.  Upgrades against later RT
    # releases will pick up --package from AddUpgradeHistory.
    if ($action eq 'upgrade' and not $has_upgrade) {
        push @args, "--package" => $name;
    }

    print "$^X @args\n";
    (system($^X, @args) == 0) or die "...returned with error: $?\n";
}

sub RTxPatch {
    my ($patch, $action, $dir) = @_;
    my @cmd;

    _rt_runtime_load();

    # Check to see if we are adding or removing a patch set
    if ( $action =~ m/\Aapply\z/xms ) {
        @cmd = ($patch, '-d', $RT::BasePath, '-p1', '-i');
    }
    else {
        @cmd = ($patch, '-R', '-d', $RT::BasePath, '-p1', '-i');
    }

    # Anonymous subroutine to apply all patches in a directory structure
    my $patch_rt = sub {

        # Next entry if not a file
        return unless -f $_;

        # Next entry if not approprate version of patch
        return unless &_required_patch($File::Find::dir, $RT::VERSION);

        if ($_ =~ m/\.patch\z/xms) {
            push @cmd, getcwd . "/$_";
            WIFEXITED(system(@cmd))
                or croak "Couldn't run: " . join(' ', @cmd) . "($?)\n";
            pop @cmd;
        }
    };

    find {wanted => $patch_rt}, $dir;
}

sub RTxPlugin {
    my ($name) = @_;

    _rt_runtime_load();
    require YAML::Tiny;
    my $data = YAML::Tiny::LoadFile('META.yml');
    my $name = $data->{name};

    my @enabled = RT->Config->Get('Plugins');
    for my $required (@{$data->{x_requires_rt_plugins} || []}) {
        next if grep {$required eq $_} @enabled;

        warn <<"EOT";

**** Warning: $name requires that the $required plugin be installed and
              enabled; it is not currently in \@Plugins.

EOT
    }
}

1;


lib/RT/Extension/CustomiseMenus.pm
use strict;
use warnings;
package RT::Extension::CustomiseMenus;

our $VERSION = '0.01';

=head1 NAME

RT-Extension-CustomiseMenus - enables easy modification of RT menus

=head1 DESCRIPTION

This extension for RT allows a administrator to easily add/remove/rename
the RT menus.

=head1 RT VERSION

Works with RT 4.4

=head1 INSTALLATION

=over

=item C<perl Makefile.PL>

=item C<make>

=item C<make install>

May need root permissions

=item Edit your F</opt/rt4/etc/RT_SiteConfig.pm>

Adding this line:

    Plugin('RT::Extension::CustomiseMenus');

=item Clear your mason cache

    rm -rf /opt/rt4/var/mason_data/obj

=item Restart your webserver

=back

=head1 CONFIGURATION

A RT menu can be manipluated by adding a simple configuration
array of hashes as shown below:

    Set($PrivilegedMenus, [
                {
                        'id'     => 'corporate',
                        'action' => 'add',
                        'path'   => 'http://bestpractical.com',
                        'title'  => 'Corporate',
                        'menu'   => 'main',
                        'after'  => 'home',
                },
                {
                        'id'     => 'wiki',
                        'action' => 'add',
                        'path'   => 'http://wiki.bestpractical.com',
                        'title'  => 'Wiki',
                        'menu'   => 'main',
                        'parent' => 'corporate',
                },
                {
                        'id'     => 'newitem',
                        'action' => 'add',
                        'path'   => '/new/thing/here',
                        'title'  => 'New Action',
                        'menu'   => 'tabs',
                        'parent' => 'actions',
                },
                {
                        'id'     => 'tools',
                        'action' => 'remove',
                        'menu'   => 'main',
                },
    ]);

    Set($SelfServiceMenus, [

    ]);

=head1 AUTHOR

Martin Wheldon E<lt>martin.wheldon@greenhills-it.co.ukE<gt>

=head1 BUGS

All bugs should be reported via email to

    L<bug-RT-Extension-CustomiseMenus@greenhills-it.co.uk|mailto:bug-RT-Extension-CustomiseMenus@greenhills-it.co.uk>

=head1 LICENSE AND COPYRIGHT

This software is Copyright (c) 2017 by Greenhills IT Ltd.

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991

=cut

use Exporter 'import';

our @EXPORT = qw/ add_menu_item remove_menu_item /;

sub load_menu
{
    my ($config) = shift;

    my $menu = {};

    # Load the appropriate submenu
    if (exists $config->{parent}) {
        my $parent = $config->{parent};

        $menu = HTML::Mason::Commands::Menu()->child($parent) if $config->{menu} eq "main";
        $menu = HTML::Mason::Commands::PageMenu()->child($parent) if $config->{menu} eq "tabs";
    }
    # Load the appropriate root menu
    else {
        $menu = HTML::Mason::Commands::Menu() if $config->{menu} eq "main";
        $menu = HTML::Mason::Commands::PageMenu() if $config->{menu} eq "tabs";
    }

    return $menu;

}

sub add_menu_item
{
    my ($config) = shift;

    my $id = $config->{id};
    my $menu = &load_menu($config);

    # FIXME: Could be add permissions here???
    #        was thinking of adding those to the anonymous hash
    # If we have a valid menu object then try to add the new menu item
    if ($menu) {
        if ( not $menu->child($id,
                                title   => HTML::Mason::Commands::loc($config->{title}),
                                path    => HTML::Mason::Commands::loc($config->{path}),
                                )) {
            $RT::Logger->error("Failed to add menu item $id");
        }
    }
    else {
        $RT::Logger->error("Failed to load parent menu of item $id");
    }

}

sub remove_menu_item
{
    my ($config) = shift;

    my $id = $config->{id};
    my $menu = &load_menu($config);

    # If we have a valid menu object then try to delete the menu item
    if ($menu) {
        if (not $menu->delete($id)) {
            $RT::Logger->error("Failed to remove menu item $id");
        }
    }
    else {
        $RT::Logger->error("Failed to load parent menu of item $id");
    }

}

1;
